### ISSUE TITLE
packer deletes VM from Esxi inventory after successful build!

### ISSUE THREAD

**Issue by ghost** [mitchellh/packer#2579]
Hi,
Packer successfully build the Centos7 image but after that it removes the vm from inventory. However, it keeps the VM in data store and I can manually add it to inventory from the data store!
Not sure, how to fix that?

Packer v0.7.5

My config----

```
{
  "builders": [
    {
      "type": "vmware-iso",
      "vm_name": "centos-rhel-7",
      "iso_url": "/home/files/CentOS-7-x86_64-Minimal-1503-01.iso",
      "iso_checksum": "d07ab3e615c66a8b2e9a50f4852e6a77",
      "iso_checksum_type": "md5",
      "disk_size": "10240",
      "disk_type_id": "thin",
      "http_directory": "~/packertemplatebuilding",
      "remote_host": "192.168.111.111",
      "remote_datastore": "SSD_60",
      "remote_username": "root",
      "remote_password": "abc@123",
      "remote_type": "esx5",
      "ssh_username": "root",
      "guest_os_type": "rhel6-64",
      "ssh_password": "abc@123",
      "ssh_port": 22,
      "ssh_wait_timeout": "20m",
      "shutdown_command": "shutdown -h now",
      "headless": "false",
      "boot_command": [
        "<tab> text ks=http://192.168.111.8/files/ksv3.cfg<enter><wait>"
      ],
      "boot_wait": "5s",
      "vmx_data": {
        "ethernet0.networkName": "VM Network",
        "config.version": 8,
        "virtualHW.version": 8,
        "ethernet0.present": "true",
        "ethernet0.virtualDev": "e1000", 
        "ethernet0.startConnected": "true",
        "ethernet0.addressType": "generated",
        "memsize": "2048",
        "numvcpus": "2",
        "cpuid.coresPerSocket": "1",
        "ide0:0.fileName": "disk.vmdk",
        "ide0:0.present": "TRUE",
        "ide0:0.redo": "",
        "scsi0:0.present": "FALSE",
        "RemoteDisplay.vnc.enabled": "TRUE",
        "RemoteDisplay.vnc.port":  "5900"

      }
    }
  ]
}

```

Packer log------

 ```

2015/08/10 20:35:07 packer-builder-vmware-iso: 2015/08/10 20:35:07 TCP connection to SSH ip/port failed: dial tcp 192.168.111.153:22: connection refused
2015/08/10 20:35:27 packer-builder-vmware-iso: 2015/08/10 20:35:27 TCP connection to SSH ip/port failed: dial tcp 192.168.111.153:22: i/o timeout
2015/08/10 20:35:32 packer-builder-vmware-iso: 2015/08/10 20:35:32 Attempting SSH connection...
2015/08/10 20:35:32 packer-builder-vmware-iso: 2015/08/10 20:35:32 reconnecting to TCP connection for SSH
2015/08/10 20:35:32 packer-builder-vmware-iso: 2015/08/10 20:35:32 handshaking with SSH
2015/08/10 20:35:32 ui: ==> vmware-iso: Connected to SSH!
==> vmware-iso: Connected to SSH!
2015/08/10 20:35:32 packer-builder-vmware-iso: 2015/08/10 20:35:32 opening new ssh session
2015/08/10 20:35:32 packer-builder-vmware-iso: 2015/08/10 20:35:32 starting remote command: vim-cmd vmsvc/tools.install 25
2015/08/10 20:35:32 packer-builder-vmware-iso: 2015/08/10 20:35:32 remote command exited with '0': vim-cmd vmsvc/tools.install 25
2015/08/10 20:35:32 packer-builder-vmware-iso: 2015/08/10 20:35:32 Running the provision hook
2015/08/10 20:35:32 ui: ==> vmware-iso: Forcibly halting virtual machine...
==> vmware-iso: Forcibly halting virtual machine...
2015/08/10 20:35:32 packer-builder-vmware-iso: 2015/08/10 20:35:32 opening new ssh session
2015/08/10 20:35:32 packer-builder-vmware-iso: 2015/08/10 20:35:32 starting remote command: vim-cmd vmsvc/power.off 25
2015/08/10 20:35:33 packer-builder-vmware-iso: 2015/08/10 20:35:33 remote command exited with '0': vim-cmd vmsvc/power.off 25
2015/08/10 20:35:33 ui:     vmware-iso: Waiting for VMware to clean up after itself...
2015/08/10 20:35:33 packer-builder-vmware-iso: 2015/08/10 20:35:33 opening new ssh session
2015/08/10 20:35:33 packer-builder-vmware-iso: 2015/08/10 20:35:33 starting remote command: ls -1p /vmfs/volumes/SSD_60/output-vmware-iso
    vmware-iso: Waiting for VMware to clean up after itself...
2015/08/10 20:35:33 packer-builder-vmware-iso: 2015/08/10 20:35:33 remote command exited with '0': ls -1p /vmfs/volumes/SSD_60/output-vmware-iso
2015/08/10 20:35:33 packer-builder-vmware-iso: 2015/08/10 20:35:33 No more lock files found. VMware is clean.
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 VM shut down.
2015/08/10 20:35:38 ui: ==> vmware-iso: Deleting unnecessary VMware files...
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 opening new ssh session
==> vmware-iso: Deleting unnecessary VMware files...
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 starting remote command: ls -1p /vmfs/volumes/SSD_60/output-vmware-iso
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 remote command exited with '0': ls -1p /vmfs/volumes/SSD_60/output-vmware-iso
2015/08/10 20:35:38 ui:     vmware-iso: Deleting: /vmfs/volumes/SSD_60/output-vmware-iso/vmware.log
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 opening new ssh session
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 starting remote command: rm /vmfs/volumes/SSD_60/output-vmware-iso/vmware.log
    vmware-iso: Deleting: /vmfs/volumes/SSD_60/output-vmware-iso/vmware.log
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 remote command exited with '0': rm /vmfs/volumes/SSD_60/output-vmware-iso/vmware.log
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 Writing VMX to: /tmp/packer-vmx905476820/centos-rhel-7.vmx
2015/08/10 20:35:38 ui: ==> vmware-iso: Cleaning VMX prior to finishing up...
==> vmware-iso: Cleaning VMX prior to finishing up...
2015/08/10 20:35:38 ui:     vmware-iso: Unmounting floppy from VMX...
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 Deleting key: floppy0.present
    vmware-iso: Unmounting floppy from VMX...
2015/08/10 20:35:38 ui:     vmware-iso: Detaching ISO from CD-ROM device...
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 Writing VMX to: /tmp/packer-vmx905476820/centos-rhel-7.vmx
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 opening new ssh session
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 Starting remote scp process:  scp -vt /vmfs/volumes/SSD_60/output-vmware-iso
    vmware-iso: Detaching ISO from CD-ROM device...
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 Started SCP session, beginning transfers...
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 Copying input data into temporary file so we can read the length
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 [DEBUG] Uploading centos-rhel-7.vmx: perms=C0644 size=2414
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 SCP session complete, closing stdin pipe.
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 Waiting for SSH session to complete.
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 scp stderr (length 35): Sink: C0644 2414 centos-rhel-7.vmx
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 opening new ssh session
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 starting remote command: vim-cmd vmsvc/reload 25
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 remote command exited with '0': vim-cmd vmsvc/reload 25
2015/08/10 20:35:38 ui: ==> vmware-iso: Compacting the disk image
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 opening new ssh session
2015/08/10 20:35:38 packer-builder-vmware-iso: 2015/08/10 20:35:38 starting remote command: vim-cmd vmsvc/power.getstate 25
==> vmware-iso: Compacting the disk image
2015/08/10 20:35:39 packer-builder-vmware-iso: 2015/08/10 20:35:39 remote command exited with '0': vim-cmd vmsvc/power.getstate 25
2015/08/10 20:35:39 ui: ==> vmware-iso: Unregistering virtual machine...
2015/08/10 20:35:39 packer-builder-vmware-iso: 2015/08/10 20:35:39 opening new ssh session
2015/08/10 20:35:39 packer-builder-vmware-iso: 2015/08/10 20:35:39 starting remote command: vim-cmd vmsvc/unregister 25
==> vmware-iso: Unregistering virtual machine...
2015/08/10 20:35:39 packer-builder-vmware-iso: 2015/08/10 20:35:39 remote command exited with '0': vim-cmd vmsvc/unregister 25
2015/08/10 20:35:39 packer-builder-vmware-iso: 2015/08/10 20:35:39 opening new ssh session
2015/08/10 20:35:39 packer-builder-vmware-iso: 2015/08/10 20:35:39 starting remote command: ls -1p /vmfs/volumes/SSD_60/output-vmware-iso
2015/08/10 20:35:39 packer-builder-vmware-iso: 2015/08/10 20:35:39 remote command exited with '0': ls -1p /vmfs/volumes/SSD_60/output-vmware-iso
2015/08/10 20:35:39 ui: Build 'vmware-iso' finished.
2015/08/10 20:35:39 Builds completed. Waiting on interrupt barrier...
2015/08/10 20:35:39 ui: 
==> Builds finished. The artifacts of successful builds are:
2015/08/10 20:35:39 machine readable: vmware-iso,artifact-count []string{"1"}
Build 'vmware-iso' finished.

==> Builds finished. The artifacts of successful builds are:
2015/08/10 20:35:39 machine readable: vmware-iso,artifact []string{"0", "builder-id", "mitchellh.vmware-esx"}
2015/08/10 20:35:39 machine readable: vmware-iso,artifact []string{"0", "id", "VM"}
2015/08/10 20:35:39 machine readable: vmware-iso,artifact []string{"0", "string", "VM files in directory: /vmfs/volumes/SSD_60/output-vmware-iso"}
2015/08/10 20:35:39 machine readable: vmware-iso,artifact []string{"0", "files-count", "6"}
2015/08/10 20:35:39 machine readable: vmware-iso,artifact []string{"0", "file", "0", "/vmfs/volumes/SSD_60/output-vmware-iso/centos-rhel-7.nvram"}
2015/08/10 20:35:39 machine readable: vmware-iso,artifact []string{"0", "file", "1", "/vmfs/volumes/SSD_60/output-vmware-iso/centos-rhel-7.vmsd"}
2015/08/10 20:35:39 machine readable: vmware-iso,artifact []string{"0", "file", "2", "/vmfs/volumes/SSD_60/output-vmware-iso/centos-rhel-7.vmx"}
2015/08/10 20:35:39 machine readable: vmware-iso,artifact []string{"0", "file", "3", "/vmfs/volumes/SSD_60/output-vmware-iso/centos-rhel-7.vmxf"}
2015/08/10 20:35:39 machine readable: vmware-iso,artifact []string{"0", "file", "4", "/vmfs/volumes/SSD_60/output-vmware-iso/disk-flat.vmdk"}
2015/08/10 20:35:39 machine readable: vmware-iso,artifact []string{"0", "file", "5", "/vmfs/volumes/SSD_60/output-vmware-iso/disk.vmdk"}
2015/08/10 20:35:39 machine readable: vmware-iso,artifact []string{"0", "end"}
2015/08/10 20:35:39 ui: --> vmware-iso: VM files in directory: /vmfs/volumes/SSD_60/output-vmware-iso
2015/08/10 20:35:39 waiting for all plugin processes to complete...
--> vmware-iso: VM files in directory: /vmfs/volumes/SSD_60/output-vmware-iso
2015/08/10 20:35:39 /root/go/bin/packer-builder-vmware-iso: plugin process exited
```

**Comment by bedecarroll** [comment]
I wish I knew Go so I could submit a PR for this as it is a pain to have to re-register the VM. I believe the code responsible is:

https://github.com/mitchellh/packer/blob/master/builder/vmware/iso/step_register.go#L46-L52

Considering this is part of the cleanup stage is this by design and not a bug? Would it be better to have an option added to control this behavior instead (though I'm not sure why people would want this)?

Thank you!

**Comment by lantrix** [comment]
Considering Packer is just using ESXi as the HyperVisor to do the build on; it seems not to really be orientated towards managing VMs after a build; and I think neither should it.

My workaround for `vmware-iso` using `"remote_type": "esx5"` is to re-register the VM with Powershell PowerCLI as a post-build step. In fact, vSphere shows the left over packer build as orphaned, so I de-register thge orphaned VM first then create a new VM from the VMX files. This stops the `New-VM` cmdlet complaining that an existing VM of the same name already exists.

```powershell
    #Get Packer config
    $PackerConfig = Get-Content -Path $configFile -raw | ConvertFrom-Json

    #Connect to vCenter
    try {
       Connect-VIServer -Server $PackerConfig.vcenter
    } catch {
        Write-Error 'Failed to connect to vCenter ' + $PackerConfig.vcenter
        exit(1)
    }

    #Re-register VMX as VM with vSphere - from files left behind by Packer Build
    $ESXHost = Get-Cluster $PackerConfig.clusterName | Get-VMHost | Where{$_.Name -eq $PackerConfig.hostName}
    $Datastore = "vmstores:\" + $PackerConfig.vcenter + "@443\" + $PackerConfig.datacenterName+ "\" + $PackerConfig.datastoreName + "\" + $PackerConfig.output_directoryName + "\"
    $VMX = $PackerConfig.vm_nameText + ".vmx"
    $vmxFile = Get-Item "${Datastore}\${VMX}"
    try {
        #De-Register Orphaned VM from Packer Build
        Get-Vm -Name $PackerConfig.vm_nameText | Remove-Inventory -Confirm:$false
        #Register VMX files as new VM
        $newVM = New-VM -VMHost $ESXHost -VMFilePath $VMXFile.DatastoreFullPath -ErrorAction Stop
    } catch {
        Write-Error 'Could not initialise VM from packer files'
        exit(1)
    }
```

FYI the `$configFile` is pointing to the path of a json file which is the same [user variables](https://www.packer.io/docs/templates/user-variables.html) I pass into the packer template (with some extras used by the packer wrapper powershell scripts), so reloading it makes sense as the post-packer process script can use the same names/paths/datastore locations etc to know where to find the VMX files and re-register the VM with vSphere.

```json
    {
        "packer_template": "PackerBaseWin2012R2-VMWare.json",
        "vcenter": "myVcenter",
        "vcenterTargetFolder": "Packer",
        "clusterName": "myCluster",
        "datacenterName": "myDataCentre",
        "hostName": "esxihost.int",
        "ESXusernameText": "root",
        "ESXpasswordText": "password",
        "user": "packer",
        "password": "password",
        "vm_nameText": "PackerTemplateWin2012R2",
        "vm_networkName": "VM Network",
        "vm_disksize": "71680",
        "vm_diskmode": "thin",
        "datastoreName": "myDatastoreName",
        "OSCustomizationSpec": "InfraAsCode-DEV",
        "output_directoryName": "PackerImage",
        "template_folder": "DSCTemplates",
        "autounattend_file": "autounattend/dev/Autounattend.xml",
        "userdata_file": "scripts/vmware-userdata.ps1",
        "ISOpath": "",
        "ISOchecksum": "9a00defab26a046045d939086df78460",
        "ISOfile": "en_windows_server_2012_r2_vl_with_update_x64_dvd_6052766.iso"
    }
```

Just for reference, the `PackerBaseWin2012R2-VMWare.json` is shown below:

```json
    {
      "variables": {
        "vcenter": "",
        "clusterName": "",
        "datacenterName": "",
        "hostName": "",
        "ESXusernameText": "",
        "ESXpasswordText": "",
        "user": "",
        "password": "",
        "vm_nameText": "",
        "vm_networkName": "",
        "vm_disksize": "",
        "vm_diskmode": "",
        "datastoreName": "",
        "output_directoryName": "",
        "autounattend_file": "",
        "userdata_file": "",
        "ISOpath": "",
        "ISOchecksum": "",
        "ISOfile": ""
      },
      "builders": [
        {
          "type": "vmware-iso",
          "vm_name": "{{user `vm_nameText`}}",
          "output_directory": "{{user `output_directoryName`}}",
          "iso_url": "{{user `ISOpath`}}",
          "iso_checksum_type": "md5",
          "iso_checksum": "{{user `ISOchecksum`}}",
          "headless": true,
          "communicator": "winrm",
          "winrm_username": "{{user `user`}}",
          "winrm_password": "{{user `password`}}",
          "winrm_timeout": "60m",
          "remote_type": "esx5",
          "remote_host": "{{user `hostName`}}",
          "remote_username": "{{user `ESXusernameText`}}",
          "remote_password": "{{user `ESXpasswordText`}}",
          "remote_datastore": "{{user `datastoreName`}}",
          "tools_upload_flavor": "windows",
          "boot_wait": "8m",
          "shutdown_timeout": "1h",
          "shutdown_command": "shutdown /s /t 10 /f /d p:4:1 /c \"Packer Shutdown\"",
          "guest_os_type": "windows8srv-64",
          "disk_size": "{{user `vm_disksize`}}",
          "disk_type_id": "{{user `vm_diskmode`}}",
          "vnc_port_min": 5800,
          "vnc_port_max": 5825,
          "floppy_files": [
            "{{user `autounattend_file`}}",
            "{{user `userdata_file`}}",
            "scripts/vmware-guest-tools.ps1"
          ],
          "vmdk_name": "{{user `vm_nameText`}}",
          "vmx_data": {
            "remotedisplay.vnc.enabled": "TRUE",
            "remotedisplay.vnc.port": "5815",
            "memsize": "2048",
            "numvcpus": "1",
            "scsi0.virtualDev": "lsisas1068",
            "annotation": "DSC-Base-Windows-{{isotime \"20060102-1504Z\"}}",
            "ethernet0.networkName": "{{user `vm_networkName`}}",
            "ethernet0.present": "TRUE",
            "ethernet0.virtualDev": "e1000e",
            "ethernet0.addressType": "vpx"
          }
        }
      ],
      "provisioners": [
        {
          "type": "powershell",
          "elevated_user": "{{user `user`}}",
          "elevated_password": "{{user `password`}}",
          "inline": ["a:\\vmware-guest-tools.ps1"]
        },
        {
          "type": "powershell",
          "elevated_user": "{{user `user`}}",
          "elevated_password": "{{user `password`}}",
          "inline": ["invoke-expression ((new-object net.webclient).DownloadString('https://chocolatey.org/install.ps1'))"]
        },
        {
        "type": "powershell",
          "elevated_user": "{{user `user`}}",
          "elevated_password": "{{user `password`}}",
          "inline": ["choco install -y powershell -pre"]
        }
      ]
    }
```

**Comment by bedecarroll** [comment]
> Considering Packer is just using ESXi as the HyperVisor to do the build on; it seems not to really be orientated towards managing VMs after a build; and I think neither should it.

Not sure I understand this? Packer doesn't seem to "manage" the VM for any builder once the box has been built.

It would just be easier to place the removal behind a "remote_register_vm" or some such variable. That way we can avoid the registering song and dance (I do it with Ansible BTW).

**Comment by lantrix** [comment]
Yeah I'm doing the song and dance as you saw. But I understood packer [just to be using](https://www.packer.io/docs/builders/vmware-iso.html#remote_type) ESXi as the hypervisor for the build vs a local vmware.

> The only value accepted for this currently is "esx5". If this is not set, a desktop product will be used.

The end goal may not to actually run a VM on the same system you build on, but maybe build another output such as a vagrant box, etc. But in your use case; yes you would want to leave it registered :)

Maybe another configuration variable for `vmware-iso` can be introduced.

**Comment by kopernikus** [comment]
I'm also very interested to have the iso-builder optionally leave the vm registered. I do realize this is somewhat debatable since the builders job is just to build and not do deployments... Anyway, here is a patch for current master:
```
diff --git a/builder/vmware/iso/builder.go b/builder/vmware/iso/builder.go
index 15ef412..77f5ac4 100755
--- a/builder/vmware/iso/builder.go
+++ b/builder/vmware/iso/builder.go
@@ -46,6 +46,7 @@ type Config struct {
        Version             string   `mapstructure:"version"`
        VMName              string   `mapstructure:"vm_name"`
        BootCommand         []string `mapstructure:"boot_command"`
+       KeepRegistered      bool     `mapstructure:"keep_registered"`
        SkipCompaction      bool     `mapstructure:"skip_compaction"`
        VMXTemplatePath     string   `mapstructure:"vmx_template_path"`
        VMXDiskTemplatePath string   `mapstructure:"vmx_disk_template_path"`
@@ -147,7 +148,6 @@ func (b *Builder) Prepare(raws ...interface{}) ([]string, error) {
        if b.config.RemotePort == 0 {
                b.config.RemotePort = 22
        }
-
        if b.config.VMXTemplatePath != "" {
                if err := b.validateVMXTemplatePath(); err != nil {
                        errs = packer.MultiErrorAppend(
diff --git a/builder/vmware/iso/step_register.go b/builder/vmware/iso/step_register.go
index 96528ee..fb6364f 100644
--- a/builder/vmware/iso/step_register.go
+++ b/builder/vmware/iso/step_register.go
@@ -41,6 +41,12 @@ func (s *StepRegister) Cleanup(state multistep.StateBag) {

        driver := state.Get("driver").(vmwcommon.Driver)
        ui := state.Get("ui").(packer.Ui)
+       config := state.Get("config").(*Config)
+
+       _, cancelled := state.GetOk(multistep.StateCancelled)
+       _, halted := state.GetOk(multistep.StateHalted)
+       if (config.KeepRegistered == true) && (!cancelled && !halted) {
+               ui.Say("Keeping virtual machine registered with ESX host (keep_registered = true)")
+               return
+       }


        if remoteDriver, ok := driver.(RemoteDriver); ok {
                if s.Format == "" {
diff --git a/builder/vmware/iso/step_register_test.go b/builder/vmware/iso/step_register_test.go
index c654f83..8697743 100644
--- a/builder/vmware/iso/step_register_test.go
+++ b/builder/vmware/iso/step_register_test.go
@@ -32,6 +32,10 @@ func TestStepRegister_remoteDriver(t *testing.T) {
        step := new(StepRegister)

        driver := new(RemoteDriverMock)
+       var config Config
+       config.KeepRegistered = false
+       state.Put("config", &config)
+
        state.Put("driver", driver)
        state.Put("vmx_path", "foo")

@@ -63,3 +67,29 @@ func TestStepRegister_remoteDriver(t *testing.T) {
                t.Fatal("should unregister proper path")
        }
 }
+func TestStepRegister_WithoutUnregister_remoteDriver(t *testing.T) {
+       state := testState(t)
+       step := new(StepRegister)
+
+       driver := new(RemoteDriverMock)
+       var config Config
+       config.KeepRegistered = true
+       state.Put("config", &config)
+
+       state.Put("driver", driver)
+       state.Put("vmx_path", "foo")
+
+       // Test the run
+       if action := step.Run(state); action != multistep.ActionContinue {
+               t.Fatalf("bad action: %#v", action)
+       }
+       if _, ok := state.GetOk("error"); ok {
+               t.Fatal("should NOT have error")
+       }
+
+       // cleanup
+       step.Cleanup(state)
+       if driver.UnregisterCalled {
+               t.Fatal("unregister should not be called")
+       }
+}
```
Another option would be a post-processor which just re-uploads the ova to the esxi server. I dont know if this is currently possible since the documentation suggests only the vmware-vmx builder makes the result available as an Artifact.

**Comment by dantran** [comment]
+1

**Comment by rhass** [comment]
@kopernikus Can you open a pull request for your patch and update it to compile against the current head of master? I think it has value, and would like to use it myself.

As a side note, the `format` option uses `ovftool` on the _local system_ to convert the image from vmdk to ofa or ovf which it does it over the wire, rather than running entirely on the ESX box. This means the VM still would need to be reuploaded and imported as a template after it gets converted.

**Comment by rhass** [comment]
@kopernikus I rebased your patch and pushed it here: https://github.com/rhass/packer/tree/chef/add-keep_registered

I hope this helps!

**Comment by kopernikus** [comment]
@rhass I'm still unsure if the  (!cancelled && !halted) part should be kept. Keeping the vm registered has IMO  two benefits:
1. makes debugging easier (e.g. if the post processor fails you can test things on the VM) 
2. make it easier to deploy VMs if you are not interested in the ovftool deployment features and templates in general

For 1) the VM should be kept on failure, for 2) this is debatable ;) We could simply keep the VM if keep_registered is set, regardless of errors, otherwise we would need another switch for 1) Thoughts?

**Comment by rhass** [comment]
@kopernikus While I agree keeping the VM makes it a lot easier to debug when a failure occurs, it would be a departure from the standard behavior in packer. As a whole -- I think this is a behavior which should be a runtime option for all of packer (upon which everyone in the community would rejoice); but obviously that is a whole different project in itself. 

Nevertheless, I think having an additional option might for preserving upon failure is the right way to do this for a couple reasons.

1) Automated packer builds which fail should not persist in the ESX/vSphere inventory unless a user specifically tells packer to keep them for debugging purposes as they could be inadvertently used for deploying production systems.

2) Packer currently doesn't appear to delete the VMX and VMDK files upon failure anyway, and seems to only remove the VM from the ESX/vSphere inventory. This being the case, a user can still debug the VM by creating a new VM from the files on the datastore, albeit with a little manual effort.

**Comment by rhass** [comment]
The PR was merged into master -- we can probably resolve/close this issue.

**Comment by mwhooker** [comment]
closing. please re-open if this is still an issue
